server/src/unifyfs_collectives.c

```diff
@@ -0,0 +1,912 @@
+
+    /* fill in input struct */
+    extbcast_request_in_t in;
+    in.root = (int32_t)glb_pmi_rank;
+    in.gfid = gfid;
+    in.num_extends = num_extents;
+    in.exttree = extent_data;
+
+    /*  */
+    extbcast_request_forward(&bcast_tree, &in);
+
+    /* free bulk data handle */
+    margo_bulk_free(extent_data);
+    free(extents);
+
+    unifyfs_tree_free(&bcast_tree);
+
+    return ret;
+}
+
+/*************************************************************************
+ * filesize
+ *************************************************************************/
+
+static int filesize_forward(const unifyfs_tree_t* broadcast_tree,
+                            filesize_in_t* in, hg_size_t* filesize)
+{
+    printf("%d: BUCKEYES %s\n", glb_pmi_rank, __func__);
+    fflush(stdout);
+
+    hg_return_t hret;
+    int ret;
+    int i;
+
+    /* get info for tree */
+    int child_count  = broadcast_tree->child_count;
+    int* child_ranks = broadcast_tree->child_ranks;
+
+    hg_size_t _filesize = 0;
+
+    ret = unifyfs_inode_get_extent_size(in->gfid, &_filesize);
+    if (ret) {
+        goto out;
+    }
+
+    printf("%d:BUCKEYES sending filesize to %d children\n",
+            glb_pmi_rank, child_count);
+
+    if (child_count > 0) {
+        /* allocate memory for request objects
+         * TODO: possibly get this from memory pool */
+        unifyfs_coll_request_t* requests = calloc(child_count,
+                                      sizeof(unifyfs_coll_request_t));
+        if (!requests) {
+            ret = ENOMEM;
+            goto out;
+        }
+
+        /* forward request down the tree */
+        for (i = 0; i < child_count; i++) {
+            /* get rank of this child */
+            int child = child_ranks[i];
+            unifyfs_coll_request_t* req = &requests[i];
+
+            printf("children[%d]: %s\n",
+                   child, glb_servers[child].margo_svr_addr_str);
+
+            /* allocate handle */
+            hret = margo_create(unifyfsd_rpc_context->svr_mid,
+                                glb_servers[child].margo_svr_addr,
+                                unifyfsd_rpc_context->rpcs.filesize_id,
+                                &req->handle);
+            assert(hret == HG_SUCCESS);
+
+            /* invoke filesize request rpc on child */
+            hret = margo_iforward(req->handle, in, &req->request);
+            assert(hret == HG_SUCCESS);
+        }
+
+        /* wait for the requests to finish */
+        filesize_out_t out;
+        for (i = 0; i < child_count; i++) {
+            /* TODO: get outputs */
+            hret = margo_wait(requests[i].request);
+
+            /* get the output of the rpc */
+            hret = margo_get_output(requests[i].handle, &out);
+
+            printf("get response: ret=%d, filesize=%lu\n",
+                    out.ret, out.filesize);
+
+            /* set return value
+             * TODO: check if we have an error and handle it */
+            ret = out.ret;
```

Comment:
On these return values from the children, I think we need to compute an OR operation or something.  This code will use the return value from the last child, but I think we want to report an error if *any* child reports an error.

Reply:
Thanks @adammoody . Somehow I haven't able to check your reviews last week. I will fix this.